File: src/main/kotlin/net/rk4z/beacon/javaExtension/EventHook.java
package net.rk4z.beacon.javaExtension;

import net.rk4z.beacon.Event;
import net.rk4z.beacon.Priority;

import java.util.function.Supplier;

/**
 * This class represents an event hook, which is a handler for a specific type of event.
 * @param <T> The type of the event that this event hook can handle.
 */
@SuppressWarnings("unused")
public class EventHook<T extends Event> {
    private final Listener handlerClass;
    private final Handler<T> handler;
    private final boolean ignoresCondition;
    private Priority priority;
    private final Supplier<Boolean> condition;
    private final Long timeout;

    /**
     * Constructs an EventHook with the given parameters.
     * @param handlerClass The listener that this event hook belongs to.
     * @param handler The function that handles the event.
     * @param ignoresCondition A boolean that indicates whether this event hook ignores its condition.
     * @param priority The priority of this event hook.
     * @param condition A supplier that returns a boolean, which is the condition for this event hook.
     * @param timeout The timeout for this event hook, in milliseconds.
     */
    public EventHook(Listener handlerClass, Handler<T> handler, boolean ignoresCondition, Priority priority, Supplier<Boolean> condition, Long timeout) {
        this.handlerClass = handlerClass;
        this.handler = handler;
        this.ignoresCondition = ignoresCondition;
        this.priority = priority;
        this.condition = condition;
        this.timeout = timeout;
    }

    /**
     * Constructs an EventHook with the given parameters and a null timeout.
     * @param handlerClass The listener that this event hook belongs to.
     * @param handler The function that handles the event.
     * @param ignoresCondition A boolean that indicates whether this event hook ignores its condition.
     * @param priority The priority of this event hook.
     * @param condition A supplier that returns a boolean, which is the condition for this event hook.
     */
    public EventHook(Listener handlerClass, Handler<T> handler, boolean ignoresCondition, Priority priority, Supplier<Boolean> condition) {
        this(handlerClass, handler, ignoresCondition, priority, condition, null);
    }

    /**
     * Returns the listener that this event hook belongs to.
     * @return The listener that this event hook belongs to.
     */
    public Listener getHandlerClass() {
        return handlerClass;
    }

    /**
     * Returns the function that handles the event.
     * @return The function that handles the event.
     */
    public Handler<T> getHandler() {
        return handler;
    }

    /**
     * Returns whether this event hook ignores its condition.
     * @return A boolean that indicates whether this event hook ignores its condition.
     */
    public boolean isIgnoresCondition() {
        return ignoresCondition;
    }

    /**
     * Returns the priority of this event hook.
     * @return The priority of this event hook.
     */
    public Priority getPriority() {
        return priority;
    }

    /**
     * Returns the condition for this event hook.
     * @return A supplier that returns a boolean, which is the condition for this event hook.
     */
    public Supplier<Boolean> getCondition() {
        return condition;
    }

    /**
     * Returns the timeout for this event hook, in milliseconds.
     * @return The timeout for this event hook, in milliseconds.
     */
    public Long getTimeout() {
        return timeout;
    }

    /**
     * Sets the priority of this event hook.
     * @param priority The new priority of this event hook.
     */
    public void setPriority(Priority priority) {
        this.priority = priority;
    }
}

File: src/main/kotlin/net/rk4z/beacon/EventHook.kt
@file:Suppress("unused", "MemberVisibilityCanBePrivate")

package net.rk4z.beacon

/**
 * Represents a hook for an event.
 *
 * @param T The type of event.
 * @property handlerClass The class of the event handler.
 * @property handler The handler function for the event.
 * @property ignoresCondition Whether the condition should be ignored.
 * @property priority The priority of the event hook.
 * @property condition An optional condition that must be met for the handler to be executed.
 * @property timeout An optional timeout for the event hook.
 */
class EventHook<T : Event>(
    val handlerClass: IEventHandler,
    val handler: Handler<T>,
    val ignoresCondition: Boolean,
    val priority: Priority = Priority.NORMAL,
    val condition: (() -> Boolean)? = null,
    val timeout: Long? = null
)

/**
 * Represents a hook for a returnable event.
 *
 * @param T The type of event.
 * @param R The type of the result returned by the handler.
 * @property handlerClass The class of the event handler.
 * @property handler The handler function for the event.
 * @property ignoresCondition Whether the condition should be ignored.
 * @property priority The priority of the event hook.
 * @property condition An optional condition that must be met for the handler to be executed.
 * @property timeout An optional timeout for the event hook.
 */
class ReturnableEventHook<T : ReturnableEvent<R>, R>(
    val handlerClass: IEventHandler,
    val handler: ReturnableHandler<T, R>,
    val ignoresCondition: Boolean,
    val priority: Priority = Priority.NORMAL,
    val condition: (() -> Boolean)? = null,
    val timeout: Long? = null
)

/**
 * Registers an event handler for a specific event type.
 *
 * @param T The type of event.
 * @param condition An optional condition that must be met for the handler to be executed.
 * @param ignoresCondition Whether the condition should be ignored.
 * @param priority The priority of the event hook.
 * @param timeout An optional timeout for the event hook.
 * @param handler The handler function for the event.
 * @throws IllegalStateException If the class does not implement IEventHandler.
 */
inline fun <reified T : Event> IEventHandler.handler(
    noinline condition: () -> Boolean = { true },
    ignoresCondition: Boolean = false,
    priority: Priority = Priority.NORMAL,
    timeout: Long? = null,
    noinline handler: (T) -> Unit
) {
    EventBus.registerEventHook(
        T::class.java,
        EventHook(
            this,
            Handler(handler),
            ignoresCondition,
            priority,
            condition,
            timeout
        )
    )
}

/**
 * Registers a returnable event handler for a specific event type.
 *
 * @param T The type of event.
 * @param R The type of the result returned by the handler.
 * @param condition An optional condition that must be met for the handler to be executed.
 * @param ignoresCondition Whether the condition should be ignored.
 * @param priority The priority of the event hook.
 * @param timeout An optional timeout for the event hook.
 * @param handler The handler function for the event.
 * @throws IllegalStateException If the class does not implement IEventHandler.
 */
inline fun <reified T : ReturnableEvent<R>, R> IEventHandler.returnableHandler(
    noinline condition: () -> Boolean = { true },
    ignoresCondition: Boolean = false,
    priority: Priority = Priority.NORMAL,
    timeout: Long? = null,
    noinline returnableHandler: (T) -> R
) {
    EventBus.registerReturnableEventHook(
        T::class.java,
        ReturnableEventHook(
            this,
            ReturnableHandler(returnableHandler),
            ignoresCondition,
            priority,
            condition,
            timeout
        )
    )
}

enum class EventProcessingType {
    /**
     * Synchronous event processing.
     */
    HANDLER_ASYNC,

    /**
     * Asynchronous event processing.
     */
    ASYNC,

    /**
     * Synchronous event processing (alias for Sync).
     */
    FULL_SYNC;

    companion object {
        fun fromString(type: String): EventProcessingType {
            return when (type.uppercase()) {
                "ASYNC" -> ASYNC
                "HANDLERASYNC" -> HANDLER_ASYNC
                "FULLSYNC" -> FULL_SYNC
                else -> throw IllegalArgumentException("Unknown EventProcessingType: $type")
            }
        }
    }
}

